text/0000-futures.md

```diff
@@ -154,31 +164,316 @@ can be dropped.
+
+Since many the ownership semantics that are required here can easily be met
+through a reference-counted `Waker` implementation, a convienence method for
+defining `Waker`s is provided, which does not require implementing a `RawWaker`
+and the associated vtable manually.
+
+This convience method is based around the `ArcWake` trait. An implementor of
+an executor can define a type which implements the `ArcWake` trait as defined
+below. The `ArcWake` type defines the associated method `into_local_waker`,
+which allows to retrieve a `LocalWaker` instance from an `Arc` of this type.
+The returned instance will guarantee that the `wake()` and `wake_local` methods
+of the type which implements `ArcWake` are called, whenever `wake()` is called
+on a `Waker` or `LocalWaker`.
 
 ```rust
 /// A way of waking up a specific task.
 ///
-/// Any task executor must provide a way of signaling that a task it owns
-/// is ready to be `poll`ed again. Executors do so by providing a wakeup handle
-/// type that implements this trait.
-pub trait Wake: Send + Sync {
+/// By implementing this trait, types that are expected to be wrapped in an `Arc`
+/// can be converted into `LocalWaker` and `Waker` objects.
+/// Those Wakers can be used to signal executors that a task it owns
+/// is ready to be `poll`ed again.
+pub trait ArcWake: Send + Sync {
     /// Indicates that the associated task is ready to make progress and should
     /// be `poll`ed.
     ///
+    /// This function can called from the thread on which the `ArcWake` was created,
+    /// as well as from any other thread.
+    ///
     /// Executors generally maintain a queue of "ready" tasks; `wake` should place
     /// the associated task onto this queue.
     fn wake(self: &Arc<Self>);
 
     /// Indicates that the associated task is ready to make progress and should be polled.
-    /// This function is like wake, but can only be called from the thread on which this
-    /// `Wake` was created.
+    /// This function is like wake, but will only be called from the thread on which this
+    /// `ArcWake` was created.
+    ///
+    /// This function is marked unsafe because callers must guarantee that
+    /// they call it only on the task on which the `ArcWake` had been created.
     ///
     /// Executors generally maintain a queue of "ready" tasks; `wake_local` should place
     /// the associated task onto this queue.
-    unsafe fn wake_local(self: &Arc<Self>)
+    unsafe fn wake_local(self: &Arc<Self>);
+
+    /// Creates a `LocalWaker` from an Arc<T>, if T implements ArcWake.
+    ///
+    /// The returned `LocalWaker` will call `wake.wake_local()` when awoken.
+    ///
+    /// The returned `LocalWaker` can be converted into a `Waker` through
+    /// it's `into_waker()` method. If `wake()` is called on this `Waker`,
+    /// the `wake()` function that is defined inside this trait will get called.
+    fn into_local_waker(wake: Arc<Self>) -> LocalWaker where Self: Sized;
```

Comment:
Why isn't this a free standing function in `std::task`? For example `fn arc_waker<W: ArcWake>(waker: Arc<W>) -> LocalWaker`. Is there any reason you decided to add it to the trait?

Reply:
Yes, that's also possible. Being directly on the trait was the initial proposal from @aturon. I am fine with whatever seems most aligned with similar concerns.
Maybe it should be on an extension trait like `ArcWakeExt`? In the discussion above and the general comments there is also some discussion around other mechanisms to create wakers from `ArcWake`. The more ways there, the less likely they should maybe be directly on the trait.

Reply:
it's currently a free-standing function, and folks found it really hard to discover. Adding it to the trait was supposed to help with that problem.

Reply:
@cramertj But wouldn't adding a link to the free standing function at the `ArcWake` documentation achieve the same result?